docs/proposals/component-registration.md

```diff
@@ -0,0 +1,294 @@
+
+**Endpoints** is a set of address:port pairs (with optional port name) that describe the providing instances of a Service. Currently only Ready addresses are included, but there is a PR to add NotReady addresses: #13778. Internal Endpoints are created and managed by the Endpoints Controller.
+
+An **Internal Service** is a Service defined by Pod Selectors. Internal Endpoints are created from Internal Services.
+
+An **External Service** is a Service without Selectors. External Endpoints must be created manually (or at least are not created by vanilla Kubernetes components).
+
+
+## Abstract
+
+The goal of this proposal is to move Kubernetes a little closer to each of the following goals:
+
+- **High Availability**
+  - Allow multiple instances of each component
+- **Extensibility**
+  - Allow optional and third party components to be added to the cluster
+- **Deployment Flexibility**
+  - Allow components to be deployed on different machines and in different availability zones
+- **Operability**
+  - Expose a filterable list of component conditions on the apiserver to simplify collection of component conditions
+  - Expose a filterable list of components on the apiserver to see which components are installed/deployed
+- **Security**
+  - Cache component conditions in Etcd (via apiserver) to reduce denial of service risk caused by network fan-out (requires continuous liveness/readiness probing)
+- **Consistency/Simplicity/Reuse**
+  - Make external services more like internal services (defined by selectors of another resource)
+  - Simplify the usage of the Endpoints API to be exclusively dynamically generated (read-only)
+  - Group the Specification (Spec) and Status of external service provider instances into a single model (Endpoints is messy combination of both Spec and Status for both internal and external addresses)
+  - Register components as external services (namespaced and labeled)
+  - Implement liveness/readiness probing in a way that can be used by all external services, not just components
+
+
+## Problems
+
+### ComponentStatuses
+
+1. The set of components is hardcoded
+2. The component addresses and ports are hardcoded (localhost, default port)
+3. The `/healthz` probe path is hardcoded
+4. Etcd is included as a component
+5. Each user request performs network amplification by fanning out requests to all components (DOS risk)
+6. Probing is done in serial (slow)
+7. Inconsistent with other registry/storage APIs and their implementations
+
+
+### Endpoints
+
+1. Endpoints doesn't distinguish between Spec and Status. In most other resources the Spec is the desired state and the Status is current (or last known) state. For more details, see the [API Conventions](../devel/api-conventions.md#spec-and-status).
+2. Endpoints is both a dynamically populated resource AND a manually populated resource. This confusion of ownership makes it hard to use, hard to describe, overly complicated, and inconsistent with other API resources.
```

Comment:
Correct.  Without an explicit "Component" API there's a fine line between non-primary component (external or internal) behavior and user behavior. I suspect it will eventually fall to role-based ACLs to determine who can do what. For right now tho it could be assumed that all components will have admin access and act on behalf of the operator. Otherwise any user would be able to add their own component.